### PR TITLE
ames, gall: more sane clog handling

### PR DESCRIPTION
**Pull request by pkova**
We were investigating https://github.com/urbit/urbit/issues/6926 today when we realised that the `+check-clog` handling in ames is absolutely insane.

Ames maintains a queue of unsent messages for every flow. Publisher ships that send a lot of subscription updates want to be able to cancel their responsibility of sending these updates if the subscriber becomes unresponsive. This functionality is handled by ames `+check-clog` which notifies gall if a peer queues up a certain number of messages or the total size of messages becomes too large (configurable, but 5 messages and 100 000 bytes by default). When gall receives the clog notification it kicks the subscriber. Note that this only applies to %boons not %pleas. This is all reasonable, but here's where things get a little out of control.

`+check-clog` always iterates over every single flow a peer has. Not only that, but `+check-clog` is run _every time a ames retry timer fires._ It doesn't even matter if we're retrying a %plea or a %boon!

This PR changes the clog handling to trigger only when sending a %boon. It also makes gall handle the clog notification for a single agent instead of kicking the subscriber from all running agents.

Committing %rumors on a local copy of ~paldev takes hours without this change, with this it takes 20 seconds.

There's still some fine tuning to do on this PR but the general approach should be significantly more sane.

### CODE REVIEW

File: pkg/arvo/sys/vane/ames.hoon [hunk 1]
```diff
@@ -3655,6 +3650,16 @@
           |=  =bone
           ^+  peer-core
           abet:(call:(abed:mi:peer-core bone) %flub ~)
+        ::
+        ++  check-clog
+          |=  [=bone =ship]
+          ^+  peer-core
+          =/  =message-pump-state  (~(got by snd.peer-state) bone)
+          ?:  ?&  !=(0 (end 0 bone))
+                  (gth ~(wyt in unsent-messages.message-pump-state) msg.cong.ames-state)
```
Comment (pkova) [inline pkg/arvo/sys/vane/ames.hoon:3659]: Do we need the byte count check as well? I could put it here but it kinda feels unnecessary.

**Comment by pkova** [comment]
We discussed this today with @yosoyubik and realized we need to correlate gall subscriptions with ames boons to make this work properly. Working on that now.

File: pkg/arvo/sys/vane/ames.hoon [hunk 2]
```diff
@@ -3655,6 +3650,16 @@
           |=  =bone
           ^+  peer-core
           abet:(call:(abed:mi:peer-core bone) %flub ~)
+        ::
+        ++  check-clog
+          |=  [=bone =ship]
```
Comment (belisarius222) [inline pkg/arvo/sys/vane/ames.hoon:3655]: we shouldn't need `ship` here, should we? we're already in the peer core

**Comment by belisarius222** [comment]
I'm thinking about how to correlate `%clog` with the `%boon`.  One thing you could try would be to have `%clog` be a per-flow note that Ames can pass to a vane that just gave it a `%boon`.  Ames would pass the `%clog` using the same duct and wire as for passing a `%plea` note, so Gall would receive it on the same duct.

Unfortunately, I don't think that works because the duct doesn't specify the request path for a `%plea` -- the fact that it doesn't has always given me a strange feeling, but I haven't been sure of what to do about it.

Maybe something dumber and more direct would work better, like Gall giving an extra piece of data in the `%boon` gift that Ames will ping-pong back immediately if it gives a `%clog`.  So it wouldn't use the duct -- rather, Gall would send something like `[publisher-agent publisher-path requesting-ship]` to Ames (maybe path-encoded, to avoid type problems) in the `%boon` gift.  Ames would include that same datum back to Gall in the `%clog`, and then Gall would use that to figure out which subscription to kill.  (You might need a subscription nonce in there too.)

This mechanism could be made optional, so Clay would leave the 'extra data' field null, signifying that it doesn't want Ames to check clog.

**Comment by yosoyubik** [comment]
> giving an extra piece of data in the %boon gift 

that was our conclusion yesterday :)

**Review by joemfb — APPROVED** [review]
this will do

**Pull request merged**